kpatch-build/kpatch-build

```diff
@@ -831,6 +834,23 @@ fi
 
 [[ -n "$OOT_MODULE" ]] || grep -q vmlinux "$SRCDIR/Module.symvers" || die "truncated $SRCDIR/Module.symvers file"
 
+while read -ra sym_line; do
+    if [[ ${#sym_line[@]} != 4 ]]; then
+        die "Malformed ${TEMPDIR}/Module.symvers file"
+    fi
+
+    sym=${sym_line[1]}
+
+    read -ra patched_sym_line <<< "$(grep "\s$sym\s" "$SRCDIR/Module.symvers")"
+    if [[ ${#patched_sym_line[@]} != 4 ]]; then
+        die "Malformed symbol entry for ${sym} in ${SRCDIR}/Module.symvers file"
+    fi
+
+    if [[ "${sym_line[*]}" != "${patched_sym_line[*]}" ]]; then
+        warn "Version disagreement for symbol $sym"
+    fi
+done < "${TEMPDIR}/Module.symvers"
+
```

Comment:
If I understand correctly, this iterates through the reference build Module.symvers file, and for each symbol then looks for a match in the patched Module.symvers... if any columns are different between the two, then it will report a version disagreement.

A few notes/questions:

1. Take a look at the symvers_read() function in lookup.c, it documents some of the various Module.symvers formats we've seen over the years. Unfortunately it can be more than 4 columns in newer kernels.

1. The symbol grep match: does that work correctly if symbol foo is listed twice, for both moduleX and moduleY?

1. Can you add the corresponding patch author guide text to the PR:  https://github.com/dynup/kpatch/issues/1105#issuecomment-630233212  We may want to update with a new section that explains what kpatch-build will report when it finds disagreeing CRCs.  Let me know if you would like me to create that based on your current PR.

1. Another approach could use modprobe to dump the resulting kpatch-patch.ko's referenced symbols and CRCs with its `--dump-modversions` option.  Comparing that short list to the Module.symvers would probably be a bit faster, but we wouldn't detect any unreferenced symbols that might have changed.  I think your approach is more complete, but just thought I'd mention an alternative.

Reply:
A few answers :) :

1. Ah thanks for pointing that out. I wrongfully assumed there would be a single format. I'll investigate a bit more on that to get a new version ready.

2. Hmm... probably not, I assumed that there would be one entry per symbol. Otherwise I don't know modules getting linked to a kernel know which entry to pick. But maybe that's a wrong assumption from my end and can try to get something a bit more robust.

3. Good point. I'll add that in the next version.

4. Yes, I guess that approach just avoids having a bad surprise at load time. However it doesn't help much in the cumulative patch approach. But it's good to know that option exists.

Thanks.

Reply:
Regarding 2. : In case of duplicate entries in the patched symvers, the current code will only consider the first entry for the symbol.

Should such a case be dealt in another way? What are the cases where we end up with duplicate entries for a same symbol in the same file? And in such a case, are both entries strictly the same? If not, does one predominates over the other?

Reply:
Re: 2.  Thinking about it this morning, perhaps it's a case that can't exist (two similarly named exported symbols), but perhaps it's allowed for different module versions.  I also wonder if export namespacing may allow this to happen.  For now, let's set this aside as paranoid and I'll poke around with trying to prove or disprove this case. : )

Reply:
Another stray thought as I was updating the author guide doc... Should we add approach (4) as a fatal error check?  The CRC changes occur in two ways:  a) referenced symbols included in --dump-modversions will result in a kpatch-patch.ko file that can't be loaded.  Perhaps we shouldn't even create such .ko files.  b) unreferenced symbols that may only result in a future problem should a new accumulative patch introduce a reference.  These can be loaded today, so maybe deserve only a warning.

Reply:
I did wonder initially whether I should only emit a warning or an error in the check I added.

I like your suggestion as it will avoid late bad surprises (and that klp module should be unusable anyway, so appart from taking space on drive I don't really see the point of producing it).

I'll try to get that working.